core/network/impl/protocols/state_protocol_impl.cpp

```diff
@@ -0,0 +1,339 @@
+            return;
+          }
+          auto &stream = stream_res.value();
+
+          SL_DEBUG(self->log_,
+                   "Established connection over {} stream with {}",
+                   self->protocol_,
+                   peer_id);
+
+          cb(std::move(stream));
+        });
+  }
+
+  void StateProtocolImpl::readRequest(std::shared_ptr<Stream> stream) {
+    auto read_writer = std::make_shared<ProtobufMessageReadWriter>(stream);
+
+    SL_DEBUG(log_,
+             "Read request from incoming {} stream with {}",
+             protocol_,
+             stream->remotePeerId().value());
+
+    read_writer->read<StateRequest>([stream, wp = weak_from_this()](
+                                        auto &&state_request_res) mutable {
+      auto self = wp.lock();
+      if (not self) {
+        stream->reset();
+        return;
+      }
+
+      if (not state_request_res.has_value()) {
+        SL_VERBOSE(self->log_,
+                   "Error at read request from incoming {} stream with {}: {}",
+                   self->protocol_,
+                   stream->remotePeerId().value(),
+                   state_request_res.error().message());
+
+        stream->reset();
+        return;
+      }
+
+      auto &state_request = state_request_res.value();
+
+      if (self->log_->level() >= log::Level::VERBOSE) {
+        std::string keys;
+        if (!state_request.start.empty()) {
+          keys = " starting with keys [";
+          for (const auto &key : state_request.start) {
+            keys += key.toHex() + ",";
+          }
+          keys.back() = ']';
+        }
+        SL_VERBOSE(
+            self->log_,
+            "State request is received from incoming {} stream with {} for "
+            "block {}{}.",
+            self->protocol_,
+            stream->remotePeerId().value(),
+            state_request.hash.toHex(),
+            keys);
+      }
+
+      auto state_response_res =
+          self->state_observer_->onStateRequest(state_request);
+
+      if (not state_response_res) {
+        SL_VERBOSE(
+            self->log_,
+            "Error at execute request from incoming {} stream with {}: {}",
+            self->protocol_,
+            stream->remotePeerId().value(),
+            state_response_res.error().message());
+
+        stream->reset();
+        return;
+      }
+      const auto &state_response = state_response_res.value();
+
+      self->writeResponse(std::move(stream), state_response);
+    });
+  }
+
+  void StateProtocolImpl::request(
+      const PeerId &peer_id,
+      StateRequest state_request,
+      std::function<void(outcome::result<StateResponse>)> &&response_handler) {
+    auto addresses_res =
+        host_.getPeerRepository().getAddressRepository().getAddresses(peer_id);
+    if (not addresses_res.has_value()) {
+      response_handler(addresses_res.as_failure());
+      return;
+    }
+
+    newOutgoingStream(
+        {peer_id, addresses_res.value()},
```

Comment:
```c++
    newOutgoingStream(
        {peer_id, std::move(addresses_res).value()},

Reply:
Again moving result. What's the trick? Done moving value